Object of class Exception could not be converted to int

```sh
$ ./bin/phpunit --filter ExampleTest
PHPUnit 9.5.27 by Sebastian Bergmann and contributors.

E                                                                   1 / 1 (100%)

Time: 00:00.023, Memory: 30.00 MB

There was 1 error:

1) CoreTest\ExampleTest::testShouldRegisterException
ErrorException: Notice: Object of class Exception could not be converted to int

tests/CoreTest/ExampleTest.php:26
```

Example of testcase:
```php
<?php

namespace CoreTest;

use PHPUnit\Framework\TestCase;
use Prophecy\Argument;
use Prophecy\PhpUnit\ProphecyTrait;
use Psr\Log\LoggerInterface;

class ExampleTest extends TestCase
{
    use ProphecyTrait;

    public function testShouldRegisterException()
    {
        $logger = $this->prophesize(LoggerInterface::class);

        $logger->error(Argument::type('string'), Argument::allOf(
            Argument::withEntry('id', 10),
            Argument::withEntry('exception', new \Exception('wrong')),
        ))
            ->shouldBeCalled();

        $logger->reveal()->error('message', [
            'id' => 10,
            'exception' => new \Exception('wrong'),
            'other' => rand(0, 1000),
        ]);
    }
}
```

The error is thrown because the `ExactValueToken` checks if the value or argument maybe a `object`, and if only one of them is, it checks for a `__toString` in the `object`.

But PHP does not convert the `object` to `string`, before comparing it to a `int|float`.

When one of them (value or argument) is a `object` and the other is a `int|float` it needs to cast the `object` to `string` before comparing OR return `false` because they do not have the same type.

I already opened a PR to fix it: #583

--- a/spec/Prophecy/Argument/Token/ExactValueTokenSpec.php
+++ b/spec/Prophecy/Argument/Token/ExactValueTokenSpec.php
@@ -3,6 +3,7 @@
 namespace spec\Prophecy\Argument\Token;
 
 use PhpSpec\ObjectBehavior;
+use Stringable;
 
 class ExactValueTokenSpec extends ObjectBehavior
 {
@@ -143,6 +144,33 @@ class ExactValueTokenSpec extends ObjectBehavior
         $idRegexExpr = '[0-9]+';
         $this->__toString()->shouldMatch(sprintf('/^%s$/', sprintf(preg_quote("$objHash"), $idRegexExpr, $idRegexExpr)));
     }
+
+    function it_scores_10_if_value_an_numeric_and_equal_to_argument_as_stringable()
+    {
+        $value = 10;
+        $argument = new ExactValueTokenC("10");
+
+        $this->beConstructedWith($value);
+        $this->scoreArgument($argument)->shouldReturn(10);
+    }
+
+    function it_does_not_scores_if_value_an_numeric_and_equal_to_argument_as_stringable()
+    {
+        $value = 10;
+        $argument = new ExactValueTokenC("example");
+
+        $this->beConstructedWith($value);
+        $this->scoreArgument($argument)->shouldReturn(false);
+    }
+
+    function it_does_not_scores_if_value_an_object_and_not_equal_to_argument_object()
+    {
+        $value = new ExactValueTokenFixtureA;
+        $argument = new ExactValueTokenC("example");
+
+        $this->beConstructedWith($value);
+        $this->scoreArgument($argument)->shouldReturn(false);
+    }
 }
 
 class ExactValueTokenFixtureA
@@ -158,5 +186,19 @@ class ExactValueTokenFixtureB extends ExactValueTokenFixtureA
     public function __construct($value)
     {
         $this->value = $value;
+    }
+}
+
+class ExactValueTokenC implements \Stringable
+{
+    public $value;
+    public function __construct($value)
+    {
+        $this->value = $value;
+    }
+
+    public function __toString()
+    {
+        return $this->value;
     }
 }

--- a/spec/Prophecy/Argument/Token/ExactValueTokenSpec.php
+++ b/spec/Prophecy/Argument/Token/ExactValueTokenSpec.php
@@ -189,7 +189,7 @@ class ExactValueTokenFixtureB extends ExactValueTokenFixtureA
     }
 }
 
-class ExactValueTokenC implements \Stringable
+class ExactValueTokenC
 {
     public $value;
     public function __construct($value)

--- a/spec/Prophecy/Argument/Token/ExactValueTokenSpec.php
+++ b/spec/Prophecy/Argument/Token/ExactValueTokenSpec.php
@@ -3,7 +3,6 @@
 namespace spec\Prophecy\Argument\Token;
 
 use PhpSpec\ObjectBehavior;
-use Stringable;
 
 class ExactValueTokenSpec extends ObjectBehavior
 {

--- a/src/Prophecy/Argument/Token/ExactValueToken.php
+++ b/src/Prophecy/Argument/Token/ExactValueToken.php
@@ -73,6 +73,10 @@ class ExactValueToken implements TokenInterface
             if (is_object($this->value) && !method_exists($this->value, '__toString')) {
                 return false;
             }
+
+            if (is_numeric($argument) xor is_numeric($this->value)) {
+                return strval($argument) == strval($this->value) ? 10 : false;
+            }
         } elseif (is_numeric($argument) && is_numeric($this->value)) {
             // noop
         } elseif (gettype($argument) !== gettype($this->value)) {

